torch.onnx.export deals incorrectly with padding dimensions for F.pad

## 🐛 Bug

When one exports a simple `nn.Module` containing `F.pad` to ONNX, and then runs that ONNX model via onnxruntime, one gets different results. The issue is that the produced .onnx file has the padding dimensions in the wrong order. 

I have verified that the bug is in `torch.onnx.export` not in onnxruntime. Firstly, I've looked at the .onnx file. Second, I've compared the outputs produced by onnxruntime and [onnx2pytorch](https://github.com/ToriML/onnx2pytorch) -- which both agree with each other, yet are different from the original PyTorch outputs.

## To Reproduce

Run the following snippet:

```import io
import numpy as np
import onnx
import onnxruntime as ort
import torch
import torch.nn.functional as F

from torch import nn


class StackTime(nn.Module):
    __constants__ = ["factor"]
    def __init__(self, factor):
        super().__init__()
        self.factor = int(factor)
    def forward(self, x, x_lens):
        r = torch.transpose(x, 0, 1)
        s = r.shape
        r = F.pad(r, [0, 0, 0, (-s[1]) % self.factor, 0, 0])
        s = r.shape
        rs = [s[0], s[1]//self.factor, s[2]*self.factor]
        r = torch.reshape(r, rs)
        x_lens = torch.ceil(x_lens.float() / self.factor).int()
        return torch.transpose(r, 0, 1), x_lens

stm = StackTime(factor=2)

max_seq_len = 7
batch = 8
hidden = 3

torch.manual_seed(42)
x = torch.rand(max_seq_len, batch, hidden)
x_lens = torch.tensor([1, 2, 3, 4, 5, 6, 7, 7])
sx, sx_lens = stm(x, x_lens)

bitstream = io.BytesIO()
torch.onnx.export(
    model=stm,
    args=(x, x_lens),
    f=bitstream,
    input_names=["x", "x_lens"],
    opset_version=11,
    dynamic_axes={"x": {0: "seq_len", 1: "batch"}, "x_lens": {0: "batch"}})
bitstream_data = bitstream.getvalue()
ort_session = ort.InferenceSession(bitstream_data)
ort_inputs = {"x": x.numpy(), "x_lens": x_lens.numpy()}
ort_outputs = ort_session.run(None, ort_inputs)
ort_sx, ort_sx_lens = ort_outputs

print(x.detach().numpy().shape)  # (7, 8, 3)
print(x_lens.detach().numpy().shape)  # (8,)
print(sx.detach().numpy().shape)  # (4, 8, 6)
print(sx_lens.detach().numpy().shape)  # (8,)
print(ort_sx.shape)  # (3, 8, 6) -- this is different
print(ort_sx_lens.shape)  # (8,)

np.testing.assert_allclose(ort_sx, sx.detach().numpy(), rtol=1e-5, atol=1e-5)
np.testing.assert_allclose(ort_sx_lens, sx_lens.detach().numpy(), rtol=1e-5, atol=1e-5)
```

## Expected behavior

The resulting output is:
```(7, 8, 3)
(8,)
(4, 8, 6)
(8,)
(3, 8, 6)
(8,)
Traceback (most recent call last):
  File "stacktimev2.py", line 59, in <module>
    np.testing.assert_allclose(ort_sx, sx.detach().numpy(), rtol=1e-5, atol=1e-5)
  File "/home/calvin/miniconda3/envs/nonideal-quant-2/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 1528, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/home/calvin/miniconda3/envs/nonideal-quant-2/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 759, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=1e-05

(shapes (3, 8, 6), (4, 8, 6) mismatch)
 x: array([[[0.882269, 0.915004, 0.382864, 0.105315, 0.269495, 0.358813],
        [0.959306, 0.390448, 0.600895, 0.199364, 0.547192, 0.00616 ],
        [0.256572, 0.793641, 0.940771, 0.951555, 0.075266, 0.886014],...
 y: array([[[0.882269, 0.915004, 0.382864, 0.105315, 0.269495, 0.358813],
        [0.959306, 0.390448, 0.600895, 0.199364, 0.547192, 0.00616 ],
        [0.256572, 0.793641, 0.940771, 0.951555, 0.075266, 0.886014],...
```

But we expected to not get this AssertionError, because the ONNX should produce output of shape (4, 8, 6) not (3, 8, 6).

## Environment

Collecting environment information...
PyTorch version: 1.7.1+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10

Python version: 3.7 (64-bit runtime)
Python platform: Linux-4.15.0-128-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: A100-PCIE-40GB
GPU 1: A100-PCIE-40GB
GPU 2: A100-PCIE-40GB
GPU 3: A100-PCIE-40GB
GPU 4: A100-PCIE-40GB
GPU 5: A100-PCIE-40GB
GPU 6: A100-PCIE-40GB
GPU 7: A100-PCIE-40GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] onnx2pytorch==0.3.0
[pip3] pytorch-ignite==0.4.4
[pip3] qtorch==0.2.0
[pip3] tf2pytorch==0.1337
[pip3] torch==1.7.1+cu110
[pip3] torchaudio==0.7.2
[pip3] torchfile==0.1.0
[pip3] torchvision==0.8.2+cu110
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] onnx2pytorch              0.3.0                     dev_0    <develop>
[conda] pytorch-ignite            0.4.4                    pypi_0    pypi
[conda] qtorch                    0.2.0                    pypi_0    pypi
[conda] tf2pytorch                0.1337                    dev_0    <develop>
[conda] torch                     1.7.1+cu110              pypi_0    pypi
[conda] torchaudio                0.7.2                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.8.2+cu110              pypi_0    pypi


## Additional context

<!-- Add any other context about the problem here. -->

--- a/test/onnx/test_pytorch_onnx_onnxruntime.py
+++ b/test/onnx/test_pytorch_onnx_onnxruntime.py
@@ -116,7 +116,6 @@ def ort_compare_with_pytorch(ort_outs, output, rtol, atol):
 
     # compare onnxruntime and PyTorch results
     assert len(outputs) == len(ort_outs), "number of outputs differ"
-
     # compare onnxruntime and PyTorch results
     [np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol) for out, ort_out in zip(outputs, ort_outs)]
 
@@ -1687,7 +1686,7 @@ class TestONNXRuntime(unittest.TestCase):
                     x.to(dtype=torch.float64) // y.to(dtype=torch.int64), x.to(dtype=torch.float64) // y.to(dtype=torch.float64), \
                     x.to(dtype=torch.int64) // y.to(dtype=torch.int64), x.to(dtype=torch.int64) // y
 
-        x = torch.randn(2, 3, 4)
+        x = torch.arange(-2, 4).reshape(2, 3, 1)
         y = torch.arange(1, 2 * 3 * 4 + 1).reshape(2, 3, 4)
         self.run_test(FloorDivModule(), (x, y))
 
@@ -1697,7 +1696,7 @@ class TestONNXRuntime(unittest.TestCase):
             def forward(self, x, y):
                 return x // 3, x // 2., x // y
 
-        x = torch.randn(2, 3, 4)
+        x = torch.arange(-2, 4).reshape(2, 3, 1)
         y = torch.randn(2, 3, 4)
         self.run_test(FloorDivModule(), (x, y))
 
@@ -1795,9 +1794,9 @@ class TestONNXRuntime(unittest.TestCase):
                 return (x.div(y, rounding_mode="floor"),
                         torch.div(x, y, rounding_mode="floor"))
 
-        modules = [TrueDivModule(), TruncDivModule()]
-        if self.opset_version >= 9:
-            modules.append(FloorDivModule())
+        modules = [TrueDivModule(), TruncDivModule(), FloorDivModule()]
+        # if self.opset_version >= 9:
+            # modules.append(FloorDivModule())
 
         x = (torch.randn(2, 3, 4) * 100).to(torch.int)
         y = torch.arange(1, 2 * 3 * 4 + 1).reshape(2, 3, 4).to(torch.int)
@@ -4232,6 +4231,13 @@ class TestONNXRuntime(unittest.TestCase):
         y = torch.randint(10, (2, 3, 5), dtype=torch.long)
         self.run_test(XorModel(), input=(x, y))
 
+    @skipIfUnsupportedMinOpsetVersion(11)  # float equal added after opset 11
+    def test_eq(self):
+        class EqualModel(torch.nn.Module):
+            def forward(self, input, other):
+                return input == other
+        self._test_compare_ops(EqualModel(), 2)
+
     def test_gt(self):
         class GreaterModel(torch.nn.Module):
             def forward(self, input, other):
@@ -5928,13 +5934,33 @@ class TestONNXRuntime(unittest.TestCase):
         y = torch.randn(1, 2, 1)
         self.run_test(RemainderModel(), (x, y))
 
+        x = torch.tensor([7, 6, -7, -6], dtype=torch.long)
+        y = torch.tensor([2], dtype=torch.long)
+        self.run_test(RemainderModel(), (x, y))
+
+        x = x.to(torch.float)
+        self.run_test(RemainderModel(), (x, y))
+
+        y = y.to(torch.float)
+        self.run_test(RemainderModel(), (x, y))
+
+        x = x.to(torch.int32)
+        self.run_test(RemainderModel(), (x, y))
+
     def test_remainder_scalar(self):
         class RemainderModel(torch.nn.Module):
+            def __init__(self, scalar=2.55):
+                super().__init__()
+                self.scalar = scalar
+
             def forward(self, input):
-                return torch.remainder(input, 2.55)
+                return torch.remainder(input, self.scalar)
 
         x = torch.randint(10, (2, 3))
         self.run_test(RemainderModel(), x)
+
+        x = torch.tensor([7, 6, -7, -6], dtype=torch.long)
+        self.run_test(RemainderModel(2), x)
 
     @skipIfUnsupportedMinOpsetVersion(10)
     def test_fmod(self):

--- a/test/onnx/test_pytorch_onnx_onnxruntime.py
+++ b/test/onnx/test_pytorch_onnx_onnxruntime.py
@@ -116,6 +116,7 @@ def ort_compare_with_pytorch(ort_outs, output, rtol, atol):
 
     # compare onnxruntime and PyTorch results
     assert len(outputs) == len(ort_outs), "number of outputs differ"
+
     # compare onnxruntime and PyTorch results
     [np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol) for out, ort_out in zip(outputs, ort_outs)]
 
@@ -1795,8 +1796,6 @@ class TestONNXRuntime(unittest.TestCase):
                         torch.div(x, y, rounding_mode="floor"))
 
         modules = [TrueDivModule(), TruncDivModule(), FloorDivModule()]
-        # if self.opset_version >= 9:
-            # modules.append(FloorDivModule())
 
         x = (torch.randn(2, 3, 4) * 100).to(torch.int)
         y = torch.arange(1, 2 * 3 * 4 + 1).reshape(2, 3, 4).to(torch.int)

--- a/torch/onnx/symbolic_helper.py
+++ b/torch/onnx/symbolic_helper.py
@@ -376,6 +376,15 @@ def _topk_helper(g, input, k, dim, largest=True, sorted=False, out=None):
         return g.op("TopK", input, k, axis_i=dim, largest_i=largest, sorted_i=sorted, outputs=2)
 
 
+def _lt_helper(g, input, other):
+    if _export_onnx_opset_version <= 8:
+        from torch.onnx.symbolic_opset8 import lt as _lt8
+        return _lt8(g, input, other)
+    else:
+        from torch.onnx.symbolic_opset9 import lt as _lt9
+        return _lt9(g, input, other)
+
+
 def _interpolate_warning(interpolate_mode):
     onnx_op = "onnx:Resize" if _export_onnx_opset_version >= 10 else "onnx:Upsample"
     warnings.warn("You are trying to export the model with " + onnx_op + " for ONNX opset version "

--- a/torch/onnx/symbolic_opset11.py
+++ b/torch/onnx/symbolic_opset11.py
@@ -394,6 +394,13 @@ def round(g, self):
     return g.op("Round", self)
 
 
+def remainder(g, input, other):
+    if sym_help._is_fp(input) or sym_help._is_fp(other):
+        from torch.onnx.symbolic_opset9 import remainder as _remainder_9
+        return _remainder_9(g, input, other)
+    return g.op("Mod", input, other, fmod_i=0)
+
+
 @parse_args("v", "v", "i", "i")
 def split(g, self, split_size_or_sizes, dim, _outputs=None):
     if not sym_help._is_split_static(split_size_or_sizes, _outputs):

--- a/torch/onnx/symbolic_opset7.py
+++ b/torch/onnx/symbolic_opset7.py
@@ -1,5 +1,4 @@
-from torch.onnx.symbolic_helper import _block_list_in_opset, parse_args
-import torch.onnx.symbolic_helper as sym_help
+from torch.onnx.symbolic_helper import _block_list_in_opset
 
 import torch.onnx.symbolic_opset9 as sym_opset9
 

--- a/torch/onnx/symbolic_opset7.py
+++ b/torch/onnx/symbolic_opset7.py
@@ -44,28 +44,5 @@ def min(g, self, dim_or_y=None, keepdim=None):
     return sym_opset9.min(g, self, dim_or_y, keepdim)
 
 
-def div(g, self, other, *args):
-    if len(args) == 0:
-        return sym_opset9.true_divide(g, self, other)
-    else:
-        return _div_rounding_mode(g, self, other, *args)
-
-
-@parse_args("v", "v", "s")
-def _div_rounding_mode(g, self, other, rounding_mode):
-    if rounding_mode == "floor":
-        return _floor_divide(g, self, other)
-    else:
-        return sym_opset9._div_rounding_mode(g, self, other, rounding_mode)
-
-
-def _floor_divide(g, self, other):
-    if sym_help._is_fp(self) or sym_help._is_fp(other):
-        out = sym_opset9.true_divide(g, self, other)
-        return g.op("Floor", out)
-    else:
-        raise RuntimeError("Integer floor division requires ONNX opset 9 or greater")
-
-
 for block_listed_op in block_listed_operators:
     vars()[block_listed_op] = _block_list_in_opset(block_listed_op)

--- a/torch/onnx/symbolic_opset8.py
+++ b/torch/onnx/symbolic_opset8.py
@@ -5,7 +5,6 @@ import torch.onnx.symbolic_opset9 as sym_opset9
 
 from torch.onnx.symbolic_helper import parse_args, _unimplemented, _block_list_in_opset, _try_get_scalar_type
 from torch.onnx.symbolic_opset9 import _cast_Float  # type: ignore[attr-defined]
-from torch.onnx.symbolic_opset7 import div  # noqa: F401
 
 import warnings
 

--- a/torch/onnx/symbolic_opset9.py
+++ b/torch/onnx/symbolic_opset9.py
@@ -164,8 +164,8 @@ def _floor_divide(g, self, other):
         # Division is negative if: self < 0 != other < 0
         zero = g.op("Constant", value_t=torch.tensor(0, dtype=torch.int64))
         negative = g.op("Xor",
-                        g.op("Less", self, zero),
-                        g.op("Less", other, zero))
+                        sym_help._lt_helper(g, self, zero),
+                        sym_help._lt_helper(g, other, zero))
 
         # For negative numbers with self % other != 0, subtract 1 to round down instead of up
         mod = g.op("Sub", self, g.op("Mul", div, other))
@@ -173,8 +173,8 @@ def _floor_divide(g, self, other):
                           g.op("Not", g.op("Equal", mod, zero)))
 
         one = g.op("Constant", value_t=torch.tensor(1, dtype=torch.int64))
-        fixup = g.op("Sub", div, one)
-        return g.op("Where", fixup_mask, fixup, div)
+        fixup = g.op("Mul", fixup_mask, one)
+        return g.op("Sub", div, fixup)
 
 
 def floor_divide(g, self, other):
@@ -190,24 +190,13 @@ def floordiv(g, self, other):
 # If only one input is a floating type, the other input is cast to its type
 # If neither input is a floating type, both inputs are cast to the default scalar type
 def true_divide(g, self, other):
-    # Case 1: both values are floating
-    # Performs div as usual
-    if sym_help._is_fp(self) and sym_help._is_fp(other):
+    # Case 1: either values are floating
+    # Performs div as usual.
+    # Implicit casting will be handled in scalar type analysis pass.
+    if sym_help._is_fp(self) or sym_help._is_fp(other):
         return g.op("Div", self, other)
 
-    # Case 2: self is floating, other is not
-    # Casts other to self's dtype
-    if sym_help._is_fp(self):
-        other = g.op("Cast", other, to_i=sym_help.cast_pytorch_to_onnx[self.type().scalarType()])
-        return g.op("Div", self, other)
-
-    # Case 3: other is floating, self is not
-    # Casts self to other's dtype
-    if sym_help._is_fp(other):
-        self = g.op("Cast", self, to_i=sym_help.cast_pytorch_to_onnx[other.type().scalarType()])
-        return g.op("Div", self, other)
-
-    # Case 4: neither is floating
+    # Case 2: neither is floating
     # Casts both inputs to the default scalar type
     scalar_type = torch.get_default_dtype()
     onnx_scalar_type = sym_help.cast_pytorch_to_onnx["Float"]
@@ -2944,9 +2933,7 @@ def meshgrid(g, tensor_list):
 
 
 def remainder(g, input, other):
-    div = g.op("Div", input, other)
-    if sym_help._is_fp(input) or sym_help._is_fp(other):
-        div = g.op("Floor", div)
+    div = _floor_divide(g, input, other)
     quo = g.op("Mul", div, other)
     return g.op("Sub", input, quo)
 

